pyfr/backends/hip/rocblas.py

```diff
@@ -92,31 +103,63 @@ def mul(self, a, b, out, alpha=1.0, beta=0.0):
 
         # α and β factors for C = α*(A*B) + β*C
         if a.dtype == np.float64:
-            rocblas_gemm = w.rocblas_dgemm
+            rtype = w.DATATYPE_F64_R
             alpha_ct, beta_ct = c_double(alpha), c_double(beta)
         else:
-            rocblas_gemm = w.rocblas_sgemm
+            rtype = w.DATATYPE_F32_R
             alpha_ct, beta_ct = c_float(alpha), c_float(beta)
 
-        # Convenience wrapper
-        def gemm(stream):
-            w.rocblas_set_stream(h, stream)
-            rocblas_gemm(h, opA, opB, m, n, k, alpha_ct, A, A.leaddim, B,
-                         B.leaddim, beta_ct, C, C.leaddim)
+        def get_gemm(algo):
```

Comment:
This is a weird structure

Reply:
It is required that gemm is a function that takes stream as an input, but to have the algo integer as the thing that is stored as you requested, we need a function that takes the algo int as an input and returns gemm as an output. Hence this weird structure.

Reply:
I think it should be possible with just `gemm` as a bare inner function?

Reply:
Another option is the capturing trick:

```
def gemm(algo=algo):
 ...
```
and define it inside the for loop.  Then you can save `gemm` as a Python function and it will always work (can just be called as `gemm()`).